Add various utility functions for LCIO objects

dudarboh opened the pull request:
BEGINRELEASENOTES
- Added a utility function to calculate Track momentum based on its track parameters and magnetic field
- Added methods to the LCRelationNavigator that extract the highest weight with an option to indicate weight encoding type ("track"/"cluster").
- Added a utility function to get an index of a provided object inside a given LCCollection
- Added a utility function to return a leading track from ReconstructedParticle in case it has multiple tracks attached.
ENDRELEASENOTES

TL;DR:
These are the utility functions i often use in my code and I feel like they might be useful for others.
Here is more detailed overview on them:

a. Track momentum: This is very often implemented by hand in many places.

1. https://github.com/iLCSoft/MarlinReco/blob/7c3b91368d28b80a6b414f6dd94e2d4166d9d1da/Analysis/TauFinder/src/PrepareRECParticles.cc#L210-L216
2. https://github.com/iLCSoft/MarlinReco/blob/39201c6fbfc257a30e4bb86c7e2eacca908a6729/Analysis/VertexChargeRecovery/src/TrackOperator.cc#L17-L28
3. https://github.com/iLCSoft/MarlinReco/blob/39201c6fbfc257a30e4bb86c7e2eacca908a6729/Analysis/TrackToRecoParticle/src/TrackToRecoParticleConverter.cc#L70-L81
4. https://github.com/iLCSoft/MarlinReco/blob/48c3c1cc1074702ba4106911bec4756f121d2ce6/Tracking/KinkFinder/src/KinkFinder.cc#L382-L392
5. https://github.com/iLCSoft/MarlinReco/blob/39201c6fbfc257a30e4bb86c7e2eacca908a6729/PFOID/src/CreatePDFs.cc#L403-L413
6. https://github.com/iLCSoft/MarlinReco/blob/39201c6fbfc257a30e4bb86c7e2eacca908a6729/Analysis/CLICPfoSelector/src/CLICPfoSelector.cc#L223-L234
7. https://github.com/iLCSoft/MarlinReco/blob/39201c6fbfc257a30e4bb86c7e2eacca908a6729/PFOID/src/PFOID.cc#L520-L530
8. https://github.com/iLCSoft/MarlinReco/blob/97966564af294b7a4306a4233b9161d0c3bf0dc3/TimeOfFlight/src/TOFUtils.cc#L32-L42
9. https://github.com/iLCSoft/MarlinReco/blob/39201c6fbfc257a30e4bb86c7e2eacca908a6729/Tracking/V0Finder/src/V0Finder.cc#L154-L188
10. https://github.com/iLCSoft/MarlinReco/blob/39201c6fbfc257a30e4bb86c7e2eacca908a6729/Analysis/FourMomentumCovMat/src/algebraImplementation.cc#L65-L75
11. https://github.com/iLCSoft/MarlinReco/blob/39201c6fbfc257a30e4bb86c7e2eacca908a6729/Analysis/VertexChargeRecovery/src/TrackOperator.cc#L17-L28
12. https://github.com/lcfiplus/LCFIPlus/blob/278d7bb54e2602c2c0cd170c00f72fe302794bcf/src/LCIOStorer.cc#L475-L487

In some cases it is done through the `HelixClass::Initialize_Canonical`, which, in principle has the same [implementation](https://github.com/iLCSoft/MarlinUtil/blob/b43aa30ebd745b02c7fbbb130b34f02782dc6def/source/src/HelixClass.cc#L88-L105), however it forces you to build a Helix object, even if you just need a momentum and you need to copy 5 track parameters, which most of the time makes 5 lines instead of one. Also Helix class needs a capital cleanup, due to some [issues](https://github.com/iLCSoft/MarlinUtil/issues/24), much makes it not a good place to extract momentum in a long term run. In my opinion it fits the best together with LCIO track.

b. LCRelationNavigator getMaxWeightObject:

This potentially helpful to avoid bugs, where people assume the objects are sorted from highest to lowest weight in the first place, e.g.: https://github.com/lcfiplus/LCFIPlus/issues/64.

Also, having an option to sort assuming track/cluster weight encoding (which is used in "RecoMCTruthLink" relations) might raise awareness of people, why weights are not normalized and prevent unwanted behaviour, if one wasn't aware of documentation [here](https://github.com/iLCSoft/MarlinReco/blob/541d61d96f6cb923042900f9284ccf0067d8fb10/Analysis/RecoMCTruthLink/include/RecoMCTruthLinker.h#L36-L43).

c. Getting index of an element inside a collection.

It is helpful if one needs to access an object from a similar collection which doesn't have relation links:
E.g. 
MCCollection        MarlinTrkTracks                      FixedTracksCollection                    OldTracksCollection        
MCParticle <->   Track (0x7f0b08) # 13   <->                                  # 13        <->                             # 13      

If there is no LCRelations written for new collections, then we need a track index (13) to compare objects one-by-one 


d. Getting a leading track from the PFO.
It might be a more consistent thing to do for PFOs with more than a single track, than just taking the first element in the array.

tmadlener commented on src/cpp/include/UTIL/LCRelationNavigator.h:
```diff
@@ -38,55 +38,75 @@ namespace UTIL {
      * @see getRelatedFromObjects
      */
-    virtual const EVENT::FloatVec & getRelatedFromWeights(EVENT::LCObject * to) const ;
+    const EVENT::FloatVec & getRelatedFromWeights(EVENT::LCObject * to) const ;
+
+    /** Object with a highest weight that the given from-object is related to.
+     *  LCObject is of type getToType().
+     */
+    const EVENT::LCObject* getRelatedToMaxWeightObject(EVENT::LCObject* from, std::string weightType) const ;
+
+    /** From-object related to the given object with a highest weight (the inverse relationship).
+     *  LCObject is of type getFromType().
+     */
+    const EVENT::LCObject* getRelatedFromMaxWeightObject(EVENT::LCObject* to, std::string weightType) const ;
+
+    /** The highest weight of the relations returned by a call to getRelatedToObjects(from).
+     * @see getRelatedToObjects
+     */
+    float getRelatedToMaxWeight(EVENT::LCObject* from, std::string weightType) const ;
+
+    /** The highest weight of the relations returned by a call to getRelatedFromObjects(to). 
+     * @see getRelatedFromObjects
+     */
+    float getRelatedFromMaxWeight(EVENT::LCObject* to, std::string weightType) const ;
```
Are the weights and the objects commonly used together? I could imagine there is a use case for having only the object, but is there also a use case for only the weight?

dudarboh replied:
I would say weights and object _must always_ be used together.

You always want to extract an object _with a certain weight_.
You might want the weight value as well to require certain cut, e.g.: accept only MCParticles which contributed >90% of the energy to this particular PFO.

I would much prefer here the use of `std::pair`, to return both object and weight in a single method, but I decided to be consistent with already existing code style, which has separate function to return weights and objects.

tmadlener replied:
In that case I would argue that having either a `std::pair` return value or even a
```cpp
struct RelationWeight {
  Event::LCObject* object;
  float weight;
};
```
(which should be usable like)
```cpp
const auto [obj, w] = getRelatedFromMaxWeigth(to, "");
```

should be the main implementation and if you want to keep the compatibility with the existing code base, you add simple wrappers around this to return only part of the result.

dudarboh replied:
I mean,
`maxWeight` or `maxWeightObject` can be used separately, so I don't think it makes sense to change them to return a `std::pair`. If you know that this object already has a max weight (because thats what method returns), you don't need a weight value already. 
In the last message, when i said `object must always be used together` exactly for the purpose to find the `maxWeight` object or `leastWeight` object, etc. If you take a simply object, ignoring its weight, it doesn't really make sense.
But this would require complete overhaul how do we store them. I wont go that deep here

dudarboh commented on src/cpp/src/UTIL/LCRelationNavigator.cc:
```diff
@@ -68,6 +69,65 @@ namespace UTIL{
+      const EVENT::LCObjectVec& objects = getRelatedToObjects(from);
+      const EVENT::FloatVec & weights = getRelatedToWeights(from);
+      if ( objects.empty() ) return nullptr;
+      std::function<float(float)> getWeight;
+      if (weightType == "track") getWeight = [](float encoding){ return (int(encoding)%10000)/1000.; };
+      else if (weightType == "cluster") getWeight = [](float encoding){ return (int(encoding)/10000)/1000.; };
+      else {
+        getWeight = [](float encoding){ return encoding; };
+      }
```
This doesn't work.

I think each lambda inside `if` returns its unique type, so in the end it returns an error, not knowing which type to assign to the final lambda:
```
error: inconsistent types ‘UTIL::LCRelationNavigator::getRelatedToMaxWeightObject(EVENT::LCObject*, const string&) const::<lambda()>::<lambda(float)>’
and                       ‘UTIL::LCRelationNavigator::getRelatedToMaxWeightObject(EVENT::LCObject*, const string&) const::<lambda()>::<lambda(float)>’ deduced for lambda return type
               return [](float encoding){ return (int(encoding)/10000)/1000.; };
```

tmadlener replied:
Ah right, forgot about the different types of the lambdas... In that case I think we go with the `std::function` (and it's small overhead), but in that case could we make the default usage more prominent, i.e.:

```suggestion
      std::function<float(float)> getWeight = [](float encoding) { return encoding };
      if (weightType == "track") getWeight = [](float encoding){ return (int(encoding)%10000)/1000.; };
      else if (weightType == "cluster") getWeight = [](float encoding){ return (int(encoding)/10000)/1000.; };
```

As a (more or less) purely academic aside (and for your information), The IIFE could be made to work with the following little trick:
```cpp
const auto Lambda = [](int a, int b) { return a < b; }; // Lambda will be some compiler named type
const auto FuncPtr = +[](int a, int b) { return a < b; } // The lambda will be implicitly converted to 
                                                         // a function pointer and FuncPtr will be a 
                                                         // bool (* const)(int, int)
```
By converting the lambda into a function pointer via the unary `operator+` we would be able to use the IIFE again. However, given the context we are doing this in, I think it is better to go with the more legible `std::function` approach.

dudarboh replied:
I think then I leave the `std::function` out and simply do e.g.: 

```suggestion
      float maxWeight = *std::max_element(weights.begin(), weights.end(), [&getWeight](float a, float b){return a < b ;};
      if (weightType == "track") maxWeight = *std::max_element(weights.begin(), weights.end(), [&getWeight](float a, float b){return (int(a)%10000)/1000. < (int(b)%10000)/1000.;});
      else if (weightType == "cluster") maxWeight = *std::max_element(weights.begin(), weights.end(), [&getWeight](float a, float b){return (int(a)/10000)/1000. < (int(b)/10000)/1000. ;});

      return maxWeight;
```

what do you think?

dudarboh commented on src/cpp/include/UTIL/ReconstructedParticleTools.h:
```diff
@@ -0,0 +1,22 @@
+    EVENT::Track* getLeadingTrack(const EVENT::ReconstructedParticle* particle, double bz){
+        const TrackVec& tracks = particle->getTracks();
+        if ( tracks.empty() ) return nullptr;
+        auto sortByMomentum = [&bz, &getTrackMomentum](const EVENT::Track* a, const EVENT::Track* b) { return getTrackMomentum(a, bz) < getTrackMomentum(b, bz) };
```
Indeed, I don't need capture of getTrackMomentum.. Not sure I understand why, though, hm.
lambda automatically captures other methods?

tmadlener replied:
Yes, in essence if you have a lambda, what the compiler "inserts" is something like this

```cpp
// [] (int a) { return a * 2; }
struct lambda_type_with_compiler_generated_name {
  int operator()(int a) { return a * 2; }
};

So all functions that are usually available in methods (or operator overloads) of a class are also available in lambdas. A similar reasoning applies to e.g. global variables

dudarboh commented:
@tmadlener 

I have moved all implementations to the `.cc` files, so compiler sees them. And I have added CLHEP dependency to the LCIO.

I squashed some of your comments fixes into existing commits, so it doesn't look like a mess, please take a look.

tmadlener commented on src/cpp/include/UTIL/LCCollectionTools.h:
```diff
@@ -0,0 +1,15 @@
+#include "EVENT/LCCollection.h"
+
+namespace UTIL{
+    /** Extract object index inside a given LCCollection.
```
Maybe add that `-1` will be used as value if no object can be found to this docstring here.

tmadlener commented on src/cpp/src/UTIL/LCRelationNavigator.cc:
```diff
@@ -68,6 +68,55 @@ namespace UTIL{
+      const auto& weights = getRelatedToWeights(from);
+      if ( objects.empty() ) return 0.;
+
+      float maxWeight = 0.;
+      if (weightType == "track") maxWeight = *std::max_element(weights.begin(), weights.end(), [](float a, float b){return (int(a)%10000)/1000. < (int(b)%10000)/1000.;});
+      else if (weightType == "cluster") maxWeight = *std::max_element(weights.begin(), weights.end(), [](float a, float b){return (int(a)/10000)/1000. < (int(b)/10000)/1000. ;});
+      else maxWeight = *std::max_element(weights.begin(), weights.end(), [](float a, float b){return a < b ;});
+      return maxWeight;
```
```suggestion
      if (weightType == "track" || weightType == "cluster") {
         return *std::max_element(weights.begin(), weights.end(), [](float a, float b){return (int(a)%10000)/1000. < (int(b)%10000)/1000.;});
      }
      return *std::max_element(weights.begin(), weights.end(), [](float a, float b){return a < b ;});
```
You can save the assignment to an intermediate value here, since you in any case will only ever hit exactly one of the cases here. And at least in the current implementation `"cluster"` and `"track"` have the same behavior.

gaede reviewed (commented):
Hi @dudarboh, we do not want to introduce a mandatory dependence on CLHEP.
As far as I have seen, you only need this for three vectors - this should be easy to solve with either a simple ThreeVector class in UTIL or using arrays...

dudarboh commented:
Hi @gaede,

Dont we have CLHEP dependencies already here?
https://github.com/iLCSoft/LCIO/blob/a7b74eb860183b4dd142c33c2ed8859d87c645fd/src/cpp/include/UTIL/LCFourVector.h#L6-L11

not sure if it works, because currently we don't have CLHEP dependency inside `CMakeLists.txt`, however it compiles, because the header file isn't used anywhere.

We also have a test for this `LCFourVector`, and `CMakeLists.txt` for test *has* a needed CLHEP dependency:

https://github.com/iLCSoft/LCIO/blob/a7b74eb860183b4dd142c33c2ed8859d87c645fd/tests/CMakeLists.txt#L296-L302


If we strongly don't want to introduce CLHEP dependency, then should we also remove `LCFourVector` with its test for the consistency?

gaede commented:
> EDIT: I missed the word "mandatory", was that the intended design that one could compile LCIO w/o CLHEP dependency, but then just couldn't use LCFourVector?

Yes, exactly, we want LCIO to compile w/o CLHEP. The `LCFourVector` is a historical leftover and it can be used in downstream code that actually has CLHEP but does not introduce a mandatory CLHEP dependency...

tmadlener commented on src/cpp/include/UTIL/ReconstructedParticleTools.h:
```diff
@@ -0,0 +1,15 @@
+    * @author Bohdan Dudar
+    * @version August 2022
+    */
+    EVENT::Track* getLeadingTrack(const EVENT::ReconstructedParticle* particle, double bz);
```
Potentially stupid question: Do we need the magnetic field in this case? From the implementation it is necessary for the call to `getMomentum` in the sorting predicate, but technically we could pass an (almost) arbitrary value to that in this case, because the magnetic field is essentially a constant scaling factor that does not change the ordering of the different tracks, right? That would in principle increase the ergonomics of using this function. (It is also possible to just give it a reasonable default value, so that it can be overridden)

tmadlener commented on src/cpp/include/UTIL/ReconstructedParticleTools.h:
```diff
@@ -0,0 +1,15 @@
+#include "EVENT/Track.h"
+
+namespace UTIL{
+    /** Extract the leading track in case of multiple tracks attached to a single ReconstructedParticle.
```
```suggestion
    /** Extract the leading track (sorted by momentum) in case of multiple tracks attached to a single ReconstructedParticle.
```

tmadlener commented on src/cpp/src/UTIL/ReconstructedParticleTools.cc:
```diff
@@ -0,0 +1,15 @@
+    EVENT::Track* getLeadingTrack(const EVENT::ReconstructedParticle* particle, double bz){
+        const EVENT::TrackVec& tracks = particle->getTracks();
+        if ( tracks.empty() ) return nullptr;
+        auto sortByMomentum = [bz](const EVENT::Track* a, const EVENT::Track* b) { return getTrackMomentum(a, bz) < getTrackMomentum(b, bz); };
```
This comparison is not the right one in this context here, I think. The `std::array::operator<` does an element wise comparison. Instead what we would want here is a comparison of the absolute momentum.

dudarboh replied:
is fixed by the commit removing the magnetic field from the arguments

dudarboh commented:
@tmadlener, how do I fix this for macOS?

tmadlener commented on src/cpp/src/UTIL/TrackTools.cc:
```diff
@@ -0,0 +1,31 @@
+        return mom;
+    }
+
+    std::array<double, 3> getTrackMomentum(const EVENT::TrackState* ts, double bz){
+        std::array<double, 3> mom;
+        double omega = ts->getOmega();
+        double phi = ts->getPhi();
+        double tanL = ts->getTanLambda();
+        double c_light = 299.792458; // mm/ns
+        double pt = (1e-6 * c_light * bz) / std::abs(omega);
+        mom[0] = pt*std::cos(phi);
+        mom[1] = pt*std::sin(phi);    
+        mom[2] = pt*tanL;
+        return mom;
+    }
```
This could be a templated function, right? On a first glance this seems to be the identical code:

something like
```cpp
template <typename TrackLikeT>
std::array<double, 3> getTrackMomentum(const TrackLikeT* track, double bz) {
  // ... implementation
}
```

Additionally, I think you could save creating an array explicitly, by doing
```cpp
  return {pt * std::cos(phi), pt * std::sin(phi), pt * tanL};
```

tmadlener replied:
The implementation will then have to move to the header

dudarboh replied:
How do I test then, whether it compiles?

tmadlener replied:
By adding a test ;)

https://github.com/iLCSoft/LCIO/tree/master/src/cpp/src/TESTS

has a `test_tracks.cc` and also a `test_trackstate.cc`. They seem to already create a few tracks and track states, so you should be able to check whether your new utilities actually behave as they should. Essentially the `MYTEST` calls in the second half are what you are looking for, they work like

```cpp
MYTEST(/*actual*/, /*expected*/, /*name of thing that is compared*/);
```
as long as there is an implementation of `operator==` for the thing that you try to compare this should work. You can in principle also add the `getLeadingTrack` to one of the other tests, if you find a reconstructed particle with multiple tracks attached.

dudarboh replied:
And I can't used `std::hypot(px, py, pz)` (since C++17)?
LCIO is still not C++17?

tmadlener commented:
> @tmadlener, how do I fix this for macOS?

You don't ;) I still have to remove the macOS workflow here, since github has removed runners with an appropriate versions. See, e.g. https://github.com/AIDASoft/run-lcg-view/pull/3

tmadlener commented on src/cpp/src/TESTS/test_tracks.cc:
```diff
@@ -337,6 +351,7 @@ int main(int /*argc*/, char** /*argv*/ ){
                         ss << " ref[" << l << "] " ;
                         MYTEST( refL[l] , float(k*(l+1)) , ss.str() ) ;
                     }
+                    
```
```suggestion
```
